{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5d791275-b89e-4eeb-b62f-f274c98b6fcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from os import listdir\n",
    "from os.path import join\n",
    "from itertools import compress\n",
    "import pandas as pd\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f26d20b6-8be8-40b4-bb27-3b64db83f0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readin_fastq(core_path,filename):\n",
    "    seqsraw = pd.read_csv(join(core_path,filename),sep=\"\\t\",header=None)  # txt files are tab-separated that is why we say that the seperator is \"tab\" represented by \"\\t\" (sep = \"\\t\")\n",
    "    seqs = seqsraw.iloc[range(1,seqsraw.shape[0],4),:] # the range function specifies the number of lines in each file and that onle every 4th line should be read -> fastq files have 4 lines per read, one with seqeunce and remaining 3 quality indicators\n",
    "    \n",
    "    return(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "88a3e7cc-b9d5-45df-9fd1-3ecac7a52816",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/project2/kribelba_1515/data/EXTRA_seq/2024_data/uniEXTRA_lib6p0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d826e181-f395-4176-b311-f68cc4656958",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = listdir(data_path)\n",
    "# files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cc407b1c-7eab-4431-98df-05e8a5d1ca77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for gDNA and mRNA separately\n",
    "files_gDNA = list(compress(files, [\"gDNA\" in i for i in files]))\n",
    "files_mRNA = list(compress(files, [\"mRNA\" in i for i in files]))\n",
    "\n",
    "# print(\"gDNA Files:\", files_gDNA)\n",
    "# print(\"mRNA Files:\", files_mRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "761558de-d114-48a5-8652-213d0059ccb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading gDNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep1_gDNA_polyA_S2_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep1_gDNA_polyA_S2_R1_001.fastq.gz ✅\n",
      "Reading gDNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep2_gDNA_S5_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep2_gDNA_S5_R1_001.fastq.gz ✅\n",
      "Reading gDNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep3_gDNA_S6_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep3_gDNA_S6_R1_001.fastq.gz ✅\n",
      "Reading gDNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep4_gDNA_S7_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep4_gDNA_S7_R1_001.fastq.gz ✅\n",
      "Reading gDNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep5_gDNA_S8_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep5_gDNA_S8_R1_001.fastq.gz ✅\n"
     ]
    }
   ],
   "source": [
    "gDNA_sequences = []\n",
    "for file in files_gDNA:\n",
    "    print(f\"Reading gDNA file: {file} ...\")\n",
    "    gDNA_sequences.append(readin_fastq(data_path, file))\n",
    "    print(f\"Finished reading {file} ✅\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "489c06bf-c748-42a8-b02f-103c2d4658f9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[                                                          0\n",
       " 1         CTCAGCATGACGGACAGACCCATGCACAAGGACCTATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCCTACAATCACAACATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCCGCACTTAAAACCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCGAAACATCTTCCCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTTTCCCGGCAAGCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 32976253  CTCAGCATGACGGACAGACCTGTTCACACTCACATATGGAAGATCC...\n",
       " 32976257  CTCAGCATGACGGACAGACCTATAAACCAAAGCATATGGAAGATCC...\n",
       " 32976261  CTCAGCATGACGGACAGACCGTACAAAACTTTCATATGGAAGATCC...\n",
       " 32976265  CTCAGCATGACGGACAGACCCCAGTCAGTCTGCATATGGAAGATCC...\n",
       " 32976269  CTCAGCATGACGGACAGACCACAGAAAGTAGTCATATGGAAGCTCC...\n",
       " \n",
       " [8244068 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCGGCACAAGTAACCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCATCAAGATCCACCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGACAGACCCATAGACATTCGCATATGGAAGATTCT...\n",
       " 13        CTCAGCATGACGGACAGACCTAGACATTAGATCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCCATGTCACATTTCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 62929225  CTCAGCATGCGGACAGACCATGCCTGAACCGCATATGGAAGATCCT...\n",
       " 62929229  CTCAGCATGACGGACAGACCAAACGACAAATCCATATGGAAGATCC...\n",
       " 62929233  CTCAGCATGACGGACCGACCCATTCTCCACTACATATGGAAGATCC...\n",
       " 62929237  CTCAGCATGACGGACAGACCATATATAAAGATCATATGGAAGATCC...\n",
       " 62929241  CTCAGCATGACGGACAGACCCGTGCCTCCCAACATATGGCAGATCC...\n",
       " \n",
       " [15732311 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCAACCAACTAGAGCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCATTCAAACAAACCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCTGATTGAGAAGCCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCCAAACGACGGTTCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCGTGAATAAGCAACATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 78523769  CTCAGCATGACGGACAGACCTCTAAACAAACGCATATGGAAGATCC...\n",
       " 78523773  CTCAGCATGACGGACAGACCGCAGTTAACTTTCATATGGAAGATCC...\n",
       " 78523777  CTCAGCATGACGGACAGACCTTTCCCGGCAAGCATATGGAAGATCC...\n",
       " 78523781  CTCAGCATGACGGACAGACCTAACACAATCTACATATGGAAGATCC...\n",
       " 78523785  CTCAGCATGACGGACAGACCATTCAATGACCTCATATGGAAGATCC...\n",
       " \n",
       " [19630947 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCAGAATGCTACTGCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCATAAAAGTAACCCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCCTACGACATCCGCGTATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCGCTACCATTCAGCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTACAAGTATGCCCGTATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 54022241  CTCAGCATGACGGACAGACCATTAAACTCATCCATATGGAAGATCC...\n",
       " 54022245  CTCAGCATGACGGACAGACCCAATTGCTATTTCATATGGAAGATCC...\n",
       " 54022249  CTCAGCATGACGGACAGACCAAGTCACAAACGCATATGGAAGATCC...\n",
       " 54022253  CTCAGCATGACGGACAGACCGAGACCCTCTAACATATGGAAGATCC...\n",
       " 54022257  CTCAGCATGACGGACAGACCTACAAAACGATACATATGGAAGATCC...\n",
       " \n",
       " [13505565 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCAAACATGACACTCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCAGGAGTAATAGGCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCTTAAACACAAGTCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCTAATCTATGAATCGTATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTGAAATCTCAAGCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 71691609  CTCAGCATGACGGACAGACCAATATAACTCAACATATGGAAGATCC...\n",
       " 71691613  CTCAGCATGACGGACAGACCCAAAAACGGGTTCATATGGAAGATCC...\n",
       " 71691617  CTCAGCATGACGGACAGACCACATTACCAAAACATATGGAAGATCC...\n",
       " 71691621  CTCAGCATGACGGACNGACCATAACTCTACATCATATGGAAGATCC...\n",
       " 71691625  CTCAGCATGACGGACAGACCTGTAAACAAGTGCATATGGAAGATCC...\n",
       " \n",
       " [17922907 rows x 1 columns]]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gDNA_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "adb8a747-ca16-4da5-a34e-d46f05ed1ba2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading mRNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep1_mRNA_polyA_S1_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep1_mRNA_polyA_S1_R1_001.fastq.gz ✅\n",
      "Reading mRNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep2_mRNA_polyA_S1_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep2_mRNA_polyA_S1_R1_001.fastq.gz ✅\n",
      "Reading mRNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep3_mRNA_polyA_S2_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep3_mRNA_polyA_S2_R1_001.fastq.gz ✅\n",
      "Reading mRNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep4_mRNA_polyA_S3_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep4_mRNA_polyA_S3_R1_001.fastq.gz ✅\n",
      "Reading mRNA file: RMCE_6p0_uniEXTRA_ofhublib4p75_rep5_mRNA_polyA_S4_R1_001.fastq.gz ...\n",
      "Finished reading RMCE_6p0_uniEXTRA_ofhublib4p75_rep5_mRNA_polyA_S4_R1_001.fastq.gz ✅\n"
     ]
    }
   ],
   "source": [
    "mRNA_sequences = []\n",
    "for file in files_mRNA:\n",
    "    print(f\"Reading mRNA file: {file} ...\")\n",
    "    mRNA_sequences.append(readin_fastq(data_path, file))\n",
    "    print(f\"Finished reading {file} ✅\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cdeb7d55-a19b-474d-aab2-0b6d85198928",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[                                                          0\n",
       " 1         CTCAGCATGACGGACAGACCTGGACCAACTGGCGTATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCATCTTTCCCTTGCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCACCCACGTTCAGCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCCGAACCAAGCATCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTCCAAGTATGTGCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 34301393  CTCAGCATGACGGACAGACCGATGTGTCAAAGCATATGGAAGATCC...\n",
       " 34301397  CTCAGCATGACGGACAGACCAACAAAACGAAACATATGGAAGATCC...\n",
       " 34301401  CTCAGCATGACGGACAGACNATCTTTCCCTTGCATATGGAAGATCC...\n",
       " 34301405  CTCAGCATGACGGACAGACCCTATGGGGCATCCTTATGGAAGATCC...\n",
       " 34301409  CTCAGCATGACGGACAGACCTCAGAAATTACGCGTATGGAAGATCC...\n",
       " \n",
       " [8575353 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCCAGCGACGAACGCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCTGATTGAGAAGCCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCACAAATTTCAGGCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCATCTATCCCTTGCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTGATTGAGAAGCCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 81682073  CTCAGCATGACGGACAGACCATCTAGTCCTTTCTTATGGAAGCTCC...\n",
       " 81682077  CTCAGCATGACGGACAGACCATCCTGCTGAATCATATGGAAGATCC...\n",
       " 81682081  CTCAGCATGACGGACAGACCCAACAATGGGGACATATGGAAGATCC...\n",
       " 81682085  CTCACATGACGGACAGACCATCTTTCCCTTGCATATGGAAGATCCT...\n",
       " 81682089  CTCAGCATGACGGACAGACCCAAACGACGGTTCATATGGAAGATCC...\n",
       " \n",
       " [20420523 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCCTGACACAGCCTCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCCGAACCAAGCATCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCAACGAAAGAGGGCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCTAACTTGCAAGACCTATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCCGGGACAGCTAGCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 39983929  CTCAGCATGACGGACAGACCCAGACGACAGTTCATATGGAAGATCC...\n",
       " 39983933  CTCAGCATGACGGACAGACCTGTCTGCAGCTGCATATGGAAGATCC...\n",
       " 39983937  CTCAGCATGACGGACAGACCCGGCGACGAACGCATATGGAAGATCC...\n",
       " 39983941  CTCAGCATGACGGACAGACCTAAGAAAAGAACCATATGGAAGATCC...\n",
       " 39983945  CTCAGCATGACGGACAGACCCGAACCAAGCATCATATGGAAGATCC...\n",
       " \n",
       " [9995987 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCATCTTTCCCTTGCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCTCATTACTCACTCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCCTCCTAAACAGTCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCCGGGACAGCTAGCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCTGTTTGTGTAGTCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 77707341  CTCAGCATGACGGACAGACCAAATAACAAACGCATATGGAAGATCC...\n",
       " 77707345  CTCAGCATGACGGACAGACCACACCTGATGATCCTATGGAAGATCC...\n",
       " 77707349  CTCAGCATGACGGACAGACCTGATTGAGAAGCCATATGGAAGATCC...\n",
       " 77707353  CTCCGCATGACGGACAGACCATCTTTCCCTTGCATATGGAAGATCC...\n",
       " 77707357  CTCAGCATGACGGACAGACCACAACTCAAACTCATATGGAAGATCC...\n",
       " \n",
       " [19426840 rows x 1 columns],\n",
       "                                                           0\n",
       " 1         CTCAGCATGACGGACAGACCCGAACCAAGCATCATATGGAAGATCC...\n",
       " 5         CTCAGCATGACGGACAGACCTTGTTGGACAAGCATATGGAAGATCC...\n",
       " 9         CTCAGCATGACGGACAGACCCGAACCAAGCATCATATGGAAGATCC...\n",
       " 13        CTCAGCATGACGGACAGACCTTTCGCTTGCGCCATATGGAAGATCC...\n",
       " 17        CTCAGCATGACGGACAGACCGGAAATGCCAGCCATATGGAAGATCC...\n",
       " ...                                                     ...\n",
       " 52387541  CTCAGCATGACGGACAGACCATCTTTCCCTTGCCTATGGAAGATCC...\n",
       " 52387545  CTCAGCATGACGGACAGACCACAATACTGATGCATATGGAAGATCC...\n",
       " 52387549  CTCAGCATGACGGACAGACCATCTTTCCCTTGCATATGGAAGATCC...\n",
       " 52387553  CTCAGCATGACGGACAGACCCAAACGACGGTTCATATGGAAGATCC...\n",
       " 52387557  CTCAGCATGACGGACAGACCCAAGATGCCGACCATATGGAAGATCC...\n",
       " \n",
       " [13096890 rows x 1 columns]]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mRNA_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "babca95d-5d39-4b9d-9e11-65d5d11ef0c8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "f = pd.read_csv(join(data_path,'RMCE_6p0_uniEXTRA_ofhublib4p75_rep1_gDNA_polyA_S2_R1_001.fastq.gz'),sep=\"\\t\",header=None)\n",
    "f = f.iloc[range(1,f.shape[0],4),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbb6e8c0-f03c-43c8-b4b6-8c830dc088cc",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# making sure that each sequence is 75 nucleotides\n",
    "count = 0\n",
    "for seq in f.iloc[:,0]:\n",
    "    if len(seq) == 75:\n",
    "        count += 1\n",
    "\n",
    "print(count == len(f.iloc[:,0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8042d5ef-4a50-4a13-8352-91336397b217",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# it seems that NOT all sequences start with a certain fixed primer of length 20\n",
    "count = 0\n",
    "count2 = 0\n",
    "for seq in f.iloc[:,0]:\n",
    "    if seq[:20] == 'CTCAGCATGACGGACAGACC':\n",
    "     count += 1\n",
    "    else:\n",
    "     count2 +=1\n",
    "print('number of sequences with the same starting sequence: ', count, ' | number of sequences with a different starting sequence: ', count2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "31ca288b-13ac-4670-9db4-dac45d90f3cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_barcode_with_flanks(sequence, start, end, le_length=8, ri_length=8):\n",
    "    barcode = sequence[start:end]\n",
    "    left_flank = sequence[start - le_length:start]\n",
    "    right_flank = sequence[end:end + ri_length]\n",
    "    return left_flank, barcode, right_flank\n",
    "\n",
    "def process_sequences(sequences, le, ri, start, end, seq_type='gDNA'):\n",
    "    sequences_dict = {}  # key: barcode, value: list of 5 counts (one per replicate)\n",
    "    \n",
    "    for replicate_num, df in enumerate(sequences):\n",
    "        print(f'Processing replicate {replicate_num + 1}...') # just for visualization\n",
    "        \n",
    "        extracted = df.iloc[:, 0].apply(lambda seq: extract_barcode_with_flanks(seq, start, end))\n",
    "        df_extracted = pd.DataFrame(extracted.tolist(), columns=['flank_l', 'BC', 'flank_r']) # a dataframe where rows are sequences, and columns are left-flanking, right-flanking, and barcode regions\n",
    "        \n",
    "        # filter sequences that match the given left and right flanks\n",
    "        df_filtered = df_extracted[(df_extracted['flank_l'] == le) & (df_extracted['flank_r'] == ri)]\n",
    "        \n",
    "        # count occurrences of each barcode\n",
    "        barcode_counts = Counter(df_filtered['BC'])\n",
    "        \n",
    "        # store counts in the dictionary\n",
    "        for barcode, count in barcode_counts.items():\n",
    "            if barcode not in sequences_dict:\n",
    "                sequences_dict[barcode] = [0] * replicate_num  # initialize with zeros for the past replicates\n",
    "            sequences_dict[barcode] += [0] * (replicate_num + 1 - len(sequences_dict[barcode]))\n",
    "            sequences_dict[barcode][replicate_num] = count  # assign count to each replicate\n",
    "    \n",
    "    # convert dict to dataframe\n",
    "    seq_data = [[seq] + counts for seq, counts in sequences_dict.items()]\n",
    "    seq_columns = ['sequence'] + [f'{seq_type}_rep{i+1}' for i in range(len(sequences))]\n",
    "    seq_df = pd.DataFrame(seq_data, columns=seq_columns)\n",
    "    \n",
    "    print(\"✅ Finished processing!\")\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4a357879-5242-4afe-b38c-345124f62553",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing replicate 1...\n",
      "Processing replicate 2...\n",
      "Processing replicate 3...\n",
      "Processing replicate 4...\n",
      "Processing replicate 5...\n",
      "✅ Finished processing!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>gDNA_rep1</th>\n",
       "      <th>gDNA_rep2</th>\n",
       "      <th>gDNA_rep3</th>\n",
       "      <th>gDNA_rep4</th>\n",
       "      <th>gDNA_rep5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTACAATCACAA</td>\n",
       "      <td>2182</td>\n",
       "      <td>1916.0</td>\n",
       "      <td>2567.0</td>\n",
       "      <td>2194.0</td>\n",
       "      <td>2070.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGCACTTAAAAC</td>\n",
       "      <td>2093</td>\n",
       "      <td>6212.0</td>\n",
       "      <td>9785.0</td>\n",
       "      <td>5734.0</td>\n",
       "      <td>8312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GAAACATCTTCC</td>\n",
       "      <td>13547</td>\n",
       "      <td>25639.0</td>\n",
       "      <td>31943.0</td>\n",
       "      <td>18795.0</td>\n",
       "      <td>26008.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTTCCCGGCAAG</td>\n",
       "      <td>14330</td>\n",
       "      <td>33127.0</td>\n",
       "      <td>38045.0</td>\n",
       "      <td>22147.0</td>\n",
       "      <td>27850.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATGTCTAAAAG</td>\n",
       "      <td>14426</td>\n",
       "      <td>30041.0</td>\n",
       "      <td>35140.0</td>\n",
       "      <td>22666.0</td>\n",
       "      <td>34493.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845674</th>\n",
       "      <td>CACTAACCACCT</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845675</th>\n",
       "      <td>CTAATAAAGAAG</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845676</th>\n",
       "      <td>AATCTTACGATA</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845677</th>\n",
       "      <td>CACCTAATAACC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845678</th>\n",
       "      <td>TAGCCTCGCATG</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>845679 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sequence  gDNA_rep1  gDNA_rep2  gDNA_rep3  gDNA_rep4  gDNA_rep5\n",
       "0       CTACAATCACAA       2182     1916.0     2567.0     2194.0     2070.0\n",
       "1       CGCACTTAAAAC       2093     6212.0     9785.0     5734.0     8312.0\n",
       "2       GAAACATCTTCC      13547    25639.0    31943.0    18795.0    26008.0\n",
       "3       TTTCCCGGCAAG      14330    33127.0    38045.0    22147.0    27850.0\n",
       "4       AATGTCTAAAAG      14426    30041.0    35140.0    22666.0    34493.0\n",
       "...              ...        ...        ...        ...        ...        ...\n",
       "845674  CACTAACCACCT          0        0.0        0.0        0.0        1.0\n",
       "845675  CTAATAAAGAAG          0        0.0        0.0        0.0        1.0\n",
       "845676  AATCTTACGATA          0        0.0        0.0        0.0        1.0\n",
       "845677  CACCTAATAACC          0        0.0        0.0        0.0        1.0\n",
       "845678  TAGCCTCGCATG          0        0.0        0.0        0.0        1.0\n",
       "\n",
       "[845679 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gDNA_df = process_sequences(gDNA_sequences, le='GACAGACC', ri='CATATGGA', start=20, end=32, seq_type='gDNA')\n",
    "gDNA_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac2801c9-c2c5-41b6-8d65-991dad2fb970",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "gDNA_df.to_csv('barcode_gDNA_counts.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f240b9d-f6ec-4ce0-bfdc-a00eb6ecc650",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing replicate 1...\n",
      "Processing replicate 2...\n",
      "Processing replicate 3...\n",
      "Processing replicate 4...\n",
      "Processing replicate 5...\n",
      "✅ Finished processing!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>mRNA_rep1</th>\n",
       "      <th>mRNA_rep2</th>\n",
       "      <th>mRNA_rep3</th>\n",
       "      <th>mRNA_rep4</th>\n",
       "      <th>mRNA_rep5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATCTTTCCCTTG</td>\n",
       "      <td>2022632</td>\n",
       "      <td>4152958.0</td>\n",
       "      <td>1734773.0</td>\n",
       "      <td>4793584.0</td>\n",
       "      <td>2671185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACCCACGTTCAG</td>\n",
       "      <td>164029</td>\n",
       "      <td>485576.0</td>\n",
       "      <td>153708.0</td>\n",
       "      <td>480350.0</td>\n",
       "      <td>195942.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGAACCAAGCAT</td>\n",
       "      <td>228066</td>\n",
       "      <td>535795.0</td>\n",
       "      <td>397328.0</td>\n",
       "      <td>506997.0</td>\n",
       "      <td>396643.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCCAAGTATGTG</td>\n",
       "      <td>3539</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10371.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAGACATCTGCA</td>\n",
       "      <td>10255</td>\n",
       "      <td>24594.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6874.0</td>\n",
       "      <td>48560.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344210</th>\n",
       "      <td>ACATTTGTCACT</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344211</th>\n",
       "      <td>TGCCCACTCCTT</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344212</th>\n",
       "      <td>TAACAATCAACG</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344213</th>\n",
       "      <td>CGTACTAACGCA</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344214</th>\n",
       "      <td>TTTACTGACTAC</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>344215 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sequence  mRNA_rep1  mRNA_rep2  mRNA_rep3  mRNA_rep4  mRNA_rep5\n",
       "0       ATCTTTCCCTTG    2022632  4152958.0  1734773.0  4793584.0  2671185.0\n",
       "1       ACCCACGTTCAG     164029   485576.0   153708.0   480350.0   195942.0\n",
       "2       CGAACCAAGCAT     228066   535795.0   397328.0   506997.0   396643.0\n",
       "3       TCCAAGTATGTG       3539        0.0        0.0    10371.0        NaN\n",
       "4       CAGACATCTGCA      10255    24594.0        0.0     6874.0    48560.0\n",
       "...              ...        ...        ...        ...        ...        ...\n",
       "344210  ACATTTGTCACT          0        0.0        0.0        0.0        1.0\n",
       "344211  TGCCCACTCCTT          0        0.0        0.0        0.0        1.0\n",
       "344212  TAACAATCAACG          0        0.0        0.0        0.0        1.0\n",
       "344213  CGTACTAACGCA          0        0.0        0.0        0.0        1.0\n",
       "344214  TTTACTGACTAC          0        0.0        0.0        0.0        1.0\n",
       "\n",
       "[344215 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mRNA_df = process_sequences(mRNA_sequences, le='GACAGACC', ri='CATATGGA', start=20, end=32, seq_type='mRNA')\n",
    "mRNA_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34aa7751-eccd-4a41-8ec1-e53bf6b5db12",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "mRNA_df.to_csv('barcode_mRNA_counts.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base ENV",
   "language": "python",
   "name": "baseenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
